folding-schemes/src/frontend/noname/mod.rs

```diff
@@ -0,0 +1,228 @@
+                        .map_err(|e| Error::JSONSerdeError(e.to_string()))?;
+                Some(json)
+            }
+        };
+
+        let ivc_inputs = serde_json::to_string(&NoNameJSONIVCInput::from(z_i.clone()))
+            .map_err(|e| Error::JSONSerdeError(e.to_string()))?;
+        let noname_inputs = NoNameIVCInputs::<F>::try_from((ivc_inputs, external_inputs_json))
+            .map_err(|_| SynthesisError::Unsatisfiable)?;
+        let (ivc_inputs, noname_external_inputs) = noname_inputs.try_into()?;
+        let noname_witness = self
+            .circuit
+            .generate_witness(ivc_inputs, noname_external_inputs)
+            .map_err(|e| Error::WitnessCalculationError(e.to_string()))?;
+
+        let mut assigned_z_i1 = vec![];
+        let z_i1_end_index = z_i.len() + 1;
+        for idx in 1..z_i1_end_index {
+            let value: BigUint = Into::into(noname_witness.witness[idx]);
+            let field_element = F::from(value);
+            assigned_z_i1.push(field_element);
+        }
+
+        Ok(assigned_z_i1)
+    }
+
+    fn generate_step_constraints(
+        &self,
+        cs: ark_relations::r1cs::ConstraintSystemRef<F>,
+        _i: usize,
+        z_i: Vec<ark_r1cs_std::fields::fp::FpVar<F>>,
+        external_inputs: Vec<ark_r1cs_std::fields::fp::FpVar<F>>,
+    ) -> Result<Vec<ark_r1cs_std::fields::fp::FpVar<F>>, ark_relations::r1cs::SynthesisError> {
+        let external_inputs_json = match !external_inputs.is_empty() {
+            false => None,
+            true => {
+                let json = serde_json::to_string(&NoNameJSONExternalInputs::from_fpvars(
+                    &external_inputs.clone(),
+                )?)
+                .map_err(|_| SynthesisError::Unsatisfiable)?;
+                Some(json)
+            }
+        };
+
+        let ivc_inputs = serde_json::to_string(&NoNameJSONIVCInput::from_fpvars(&z_i)?)
+            .map_err(|_| SynthesisError::Unsatisfiable)?;
+        let noname_inputs = NoNameIVCInputs::<F>::try_from((ivc_inputs, external_inputs_json))
+            .map_err(|_| SynthesisError::Unsatisfiable)?;
+        let (ivc_inputs, noname_external_inputs) = noname_inputs
+            .try_into()
+            .map_err(|_| SynthesisError::Unsatisfiable)?;
+        let noname_witness = self
+            .circuit
+            .generate_witness(ivc_inputs, noname_external_inputs)
+            .map_err(|_| SynthesisError::Unsatisfiable)?;
+
+        let mut assigned_z_i1 = vec![];
+        let z_i1_end_index = z_i.len() + 1;
+        for idx in 1..z_i1_end_index {
+            // the assigned zi1 is of the same size than the initial zi and is located in the
+            // output of the witness vector
+            // we prefer to assign z_i1 here since (1) we have to return it, (2) we cant return
+            // anything with the `generate_constraints` method used below
+            let value: BigUint = Into::into(noname_witness.witness[idx]);
```

Comment:
I assume we need the `BigUint` to be able to do from_string or similar no?
Have you checked if arkworks already has a way tio do so? Such that we don't need to deal with the BigUint?

Reply:
Also, why the hell do they use String instead of bytes? At the end they are the same. But is much simpler to play with byte types anyway.

Do you happen to know?